### PR TITLE
fix getting mount points when passing a path to the files:scan command

### LINKED ISSUE #17067: files:scan does not scan external storage using the --path option
This issue were discussed in #14507 already.

### Steps to reproduce
1. use an external storage provider like Dropbox
2. execute the files:scan command on the external folder:
`./occ files:scan --path /user/files/oc/Dropbox`
3. see the command line output, the external storage folders are not scanned

### Expected behaviour
The external storage folders should be scanned. The external storage folders are scanned if you use the --all option but then every time the whole storage has to be scanned what needs more resources.

### Actual behaviour
The external storage folders are not scanned.

### Server configuration
**Operating system**: Ubuntu 12.04.1 LTS, Kernel: 2.6.32-042stab092.3
**Web server:** Apache/2.2.22 (Ubuntu)
**Database:** mysqld 5.5.43-0ubuntu0.12.04.1
**PHP version:** 5.4.42-1+deb.sury.org~precise+1

**ownCloud version:** 8.0.4 (stable) 

**Updated from an older ownCloud or fresh install:** updated several times

**List of activated apps:**
Calendar, Contacts, News, Roundcube, External Storage Support, PDF Viewer, Share Files, Updater
**The content of config/config.php:**

<?php
$CONFIG = array (
  'instanceid' => '51eecfbd4d1a0',
  'passwordsalt' => 'salt',
  'datadirectory' => '/var/www/example.com/cloud/data',
  'dbtype' => 'mysql',
  'version' => '8.0.4.2',
  'dbname' => 'dbname',
  'dbhost' => 'localhost',
  'dbtableprefix' => 'oc_',
  'dbuser' => 'dbuser',
  'dbpassword' => 'dbpassword',
  'installed' => true,
  'maxZipInputSize' => 0,
  'allowZipDownload' => true,
  'loglevel' => 0,
  'theme' => '',
  'maintenance' => false,
  'forcessl' => true,
  'trashbin_retention_obligation' => '1',
  'trusted_domains' => 
  array (
    0 => 'cloud.example.com',
  ),
  'mail_smtpmode' => 'php',
  'mail_from_address' => 'mail',
  'mail_domain' => 'example.com',
  'appcodechecker' => false,
  'secret' => 'xxx',
);

**Are you using external storage, if yes which one:** local/smb/sftp/...
Dropbox
**Are you using encryption:** yes/no
no
**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
no

I don't know if this issue affects also other external storages than Dropbox but i would expect this as a general problem with the --path option from files:scan.

### PR DESCRIPTION
**Pull request by icewind1991**
The path is already absolute.

Fixes #17067

cc @DeepDiver1975 @PVince81

### CODE REVIEW

**Comment by PVince81** [comment]
@icewind1991 will this have any side effects ?

I'm wondering why we didn't find this earlier, is this function seldom used ?

Please add a unit test.

**Comment by icewind1991** [comment]
> will this have any side effects 

You'll have to use absolute path when using the undocumented little known feature of triggering the scanner from js.

**Comment by PVince81** [comment]
So this means JS needs fixing too ?
Because the scanning still seems to be triggered from there sometimes.

**Comment by PVince81** [comment]
Tested by @matzex here https://github.com/owncloud/core/issues/17067#issuecomment-114082714, counting as thumbs up :+1:

**Comment by icewind1991** [comment]
added unit test

File: tests/lib/files/utils/scanner.php [hunk 1]
```diff
@@ -94,6 +96,41 @@ public function testReuseExistingFile() {
 		$this->assertEquals($old, $new);
 	}
 
+	public function testScanSubMount() {
+		$uid = uniqid();
```
Comment (nickvergessen) [inline tests/lib/files/utils/scanner.php]: `$this->getUniqueID()` please

Comment (icewind1991) [inline tests/lib/files/utils/scanner.php]: fixed

**Comment by icewind1991** [comment]
@owncloud-bot retest this please

File: lib/private/files/config/mountprovidercollection.php [hunk 2]
```diff
@@ -57,8 +57,11 @@ public function __construct(IStorageFactory $loader) {
 	 */
 	public function getMountsForUser(IUser $user) {
 		$loader = $this->loader;
-		return array_reduce($this->providers, function ($mounts, IMountProvider $provider) use ($user, $loader) {
-			return array_merge($mounts, $provider->getMountsForUser($user, $loader));
+		$mounts = array_map(function (IMountProvider $provider) use ($user, $loader) {
+			return $provider->getMountsForUser($user, $loader);
+		}, $this->providers);
+		return array_reduce($mounts, function ($mounts, $providerMounts) {
+			return array_merge($mounts, $providerMounts);
```
Comment (PVince81) [inline lib/private/files/config/mountprovidercollection.php]: @icewind1991 did you change this to make it faster ?

Comment (icewind1991) [inline lib/private/files/config/mountprovidercollection.php]: just more readable and easier to debug, logic is the same

Comment (PVince81) [inline lib/private/files/config/mountprovidercollection.php]: fine by me, assuming this is properly covered by the unit test

**Comment by PVince81** [comment]
:+1:

**Comment by PVince81** [comment]
```
PHP Fatal error:  Call to a member function deleteAll() on null in /var/jenkins/workspace/pull-request-analyser-ng-simple/label/SLAVE/apps/files_sharing/tests/unsharechildren.php on line 57
09:44:09 PHP Stack trace:
09:44:09 PHP   1. {main}() /usr/local/bin/phpunit:0
09:44:09 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:537
09:44:09 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:105
09:44:09 PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:153
09:44:09 PHP   5. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php:406
09:44:09 PHP   6. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
09:44:09 PHP   7. PHPUnit_Framework_TestCase->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
09:44:09 PHP   8. PHPUnit_Framework_TestResult->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:699
09:44:09 PHP   9. PHP_Invoker->invoke() phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php:607
09:44:09 PHP  10. call_user_func_array:{phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93}() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
09:44:09 PHP  11. PHPUnit_Framework_TestCase->runBare() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
09:44:09 PHP  12. OCA\Files_sharing\Tests\UnshareChildren->tearDown() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:774
09:44:09 PHP Warning:  Unknown: Cannot destroy the zip context in Unknown on line 0
09:44:09 PHP Stack trace:
09:44:09 PHP   1. {main}() /usr/local/bin/phpunit:0
09:44:09 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:537
09:44:09 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:105
09:44:09 PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:153
09:44:09 PHP   5. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php:406
09:44:09 PHP   6. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
09:44:09 PHP   7. PHPUnit_Framework_TestCase->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
09:44:09 PHP   8. PHPUnit_Framework_TestResult->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:699
09:44:09 PHP   9. PHP_Invoker->invoke() phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php:607
09:44:09 PHP  10. call_user_func_array:{phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93}() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
09:44:09 PHP  11. PHPUnit_Framework_TestCase->runBare() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
09:44:09 PHP  12. OCA\Files_sharing\Tests\UnshareChildren->tearDown() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:774
09:44:09 PHP Warning:  Unknown: Cannot destroy the zip context in Unknown on line 0
09:44:09 PHP Stack trace:
09:44:09 PHP   1. {main}() /usr/local/bin/phpunit:0
09:44:09 PHP   2. PHPUnit_TextUI_Command::main() /usr/local/bin/phpunit:537
09:44:09 PHP   3. PHPUnit_TextUI_Command->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:105
09:44:09 PHP   4. PHPUnit_TextUI_TestRunner->doRun() phar:///usr/local/bin/phpunit/phpunit/TextUI/Command.php:153
09:44:09 PHP   5. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/TextUI/TestRunner.php:406
09:44:09 PHP   6. PHPUnit_Framework_TestSuite->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
09:44:09 PHP   7. PHPUnit_Framework_TestCase->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestSuite.php:722
09:44:09 PHP   8. PHPUnit_Framework_TestResult->run() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:699
09:44:09 PHP   9. PHP_Invoker->invoke() phar:///usr/local/bin/phpunit/phpunit/Framework/TestResult.php:607
09:44:09 PHP  10. call_user_func_array:{phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93}() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
09:44:09 PHP  11. PHPUnit_Framework_TestCase->runBare() phar:///usr/local/bin/phpunit/php-invoker/Invoker.php:93
09:44:09 PHP  12. OCA\Files_sharing\Tests\UnshareChildren->tearDown() phar:///usr/local/bin/phpunit/phpunit/Framework/TestCase.php:774
09:44:09 Build step 'Execute shell' marked build as failure
09:44:10 Recording test results
09:44:10 Setting commit status on GitHub for https://github.com/owncloud/core/commit/609f044e30fbfddfde6f59380c7b32eb6d7b92e1
09:44:11 Finished: FAILURE
```

related?

File: tests/lib/files/utils/scanner.php [hunk 3]
```diff
@@ -94,6 +96,41 @@ public function testReuseExistingFile() {
 		$this->assertEquals($old, $new);
 	}
 
+	public function testScanSubMount() {
+		$uid = $this->getUniqueID();
+		$userBackend = new \OC_User_Dummy();
+		$userBackend->createUser($uid, 'test');
+		\OC::$server->getUserManager()->registerBackend($userBackend);
```
Comment (DeepDiver1975) [inline tests/lib/files/utils/scanner.php]: missing unregister? should be in setup/teardown

**Comment by icewind1991** [comment]
@owncloud-bot retest this pelase

**Comment by icewind1991** [comment]
jenkins pr: #17228

**Comment by icewind1991** [comment]
Tests passed here: https://github.com/owncloud/core/pull/17228#issuecomment-116680859

**Comment by MorrisJobke** [comment]
Jenkins: #17309 (because the linked one has different sha sums)

**Comment by MorrisJobke** [comment]
Tests passed: https://github.com/owncloud/core/pull/17309#issuecomment-117973732

**Pull request merged**